keel-core/src/main/kotlin/com/netflix/spinnaker/keel/constraints/ConstraintEvaluator.kt

```diff
@@ -62,11 +66,10 @@ abstract class StatefulConstraintEvaluator<T : Constraint> : ConstraintEvaluator
         artifactVersion = version,
         type = constraint.type,
         status = ConstraintStatus.PENDING
-      )
-        .also {
-          deliveryConfigRepository.storeConstraintState(it)
-          // TODO: Emit an event here?
-        }
+      ).also {
+        deliveryConfigRepository.storeConstraintState(it)
+        eventPublisher.publishEvent(ConstraintStateChanged(targetEnvironment, constraint, null, it))
```

Comment:
Somewhat unrelated, but this may help with publishing a user-visible resource history event when constraints approve/reject an artifact. Cc: @asher 

P.S. It increasingly feels to me that we'll need other event types beyond "resource events" (events associated with artifacts, environments, etc.) that would be useful to expose in the UI.

Reply:
💯